### DISCUSSION TITLE
Support other Templating Engines like React

### DISCUSSION THREAD

**Discussion by matthewmueller** [livebud/bud#8] [Ideas]
Right now Bud ships with out-of-the-box Svelte support. It's also designed to support React.

I picked Svelte because it already has solutions for scoped CSS and dynamically changing the `<head>`. I'm tempted to wait on introducing React support until we have both of these features. Something like including Next.js's `styled-jsx` + `<Head/>`. Is this worth waiting for or does it not matter?

Admittedly while Svelte is growing rapidly, React is the standard here and we need to meet people where they are.

I'd also like to support 2 other templating languages: 
- plain ole' HTML
- Go's [html/template](https://pkg.go.dev/html/template)

Which probably wouldn't ship with scoped CSS support (or should they?). That make me wonder what "baseline" of support I should provide and starting with basic React, then adding scoped CSS + dynamic head later would be desired.

**Update:** I just wrote a long brain dump of all my thoughts on current view support and ideas for the future. Please have a read and let me know what you think! https://github.com/livebud/bud/discussions/188#discussioncomment-3152348

**Comment by austincollinpena** [comment]
Is there a mailing list I can sign up for so I can get updates on the project? I subscribed to the github releases too :)

**Reply by matthewmueller** [reply]
Not yet, this blew up more than I anticipated! Can you send me a quick email to hi@livebud.com? I'll add you to a list manually for now :)

**Reply by amanangira** [reply]
Mee too please, email sent.

**Reply by ManoloTonto1** [reply]
Add me as well please!! this project is something I actually wanted to start myself. Looking forward to also contributing to it!

**Reply by matthewmueller** [reply]
Will do, thanks for your interest! Bud is going through a bit of a transition at the moment, but looking forward to sharing an update soon!

**Comment by stefanos82** [comment]
First of all, congrats for the project; it looks awesome and quite promising!

Since you have mentioned the possible plan to support React, logically the next thing would be to support Vue as well.

Sounds fair, right?

Now a naive question here: how difficult would it be to go full vanilla JS instead of depending on third parties and such, whenever is necessary that is?

Also, about using Go's `html/template`, yes please! That would be really cool, especially for those users who use Hugo; they can bring in their existing experience without worrying they need to learn something new.

Edit: Seems like @josebalius asked the same thing as I in #23.

**Reply by matthewmueller** [reply]
> Sounds fair, right?

Definitely fair, but I might need to be convinced on this one 😅. I haven't played with Vue yet, but just through internet osmosis, my current feeling is that it's a blend between React and Svelte. I'm not sure what it brings to bear on it's own. I'm pretty sure my view is wrong because it seems like a lot of people like Vue so I hope to be proven wrong!

> Also, about using Go's html/template, yes please! That would be really cool, especially for those users who use Hugo; they can bring in their existing experience without worrying they need to learn something new.

Yes, I actually already had support for this but removed it after adding Svelte support (and prepping React support). It would have required a large refactor that I wasn't prepared to do at that time. I'm hoping to get it added back in soon!

> Edit: Seems like @josebalius asked the same thing as I in https://github.com/livebud/bud/issues/23.

Answered!

**Reply by matthewp** [reply]
@matthewmueller would it be possible to have a concept of "renderers" through the plugin API? Maybe it would map a file extension to a function that implements whatever template API it is that Bud needs? That way you could support any frontend framework via plugins. I work on [Astro](https://github.com/withastro/astro) and this is what we do.

Either way I do think including React by default makes a lot of sense for popularity reasons.

**Reply by matthewmueller** [reply]
Hey @matthewp, nice to see you here! 

Bud indeed has a rudimentary plugin system. I took it *just* far enough to know I could render both Svelte and React. It sounds like I should be studying Astro for a more complete solution. Could you point me in the right direction where I can learn more?

**Reply by matthewmueller** [reply]
@stefanos82 FYI, I just created a discussion for Vue support: https://github.com/livebud/bud/discussions/39

**Reply by one-aalam** [reply]
Hey @matthewmueller w.rt learning Astro, besides the [official docs](https://docs.astro.build/en/getting-started/), you can find a lot of good resources [here](https://github.com/one-aalam/awesome-astro). I'm loving how seamlessly Svelte fits as the view layer in the overall scheme of things in Bud, but _renderers_ are an amazing concept which Astro pioneered in a way, and definitely seems like a great way to let the developers bring their own template engine. 

Astro's latest releases call it just [integrations](https://docs.astro.build/en/guides/integrations-guide/) now, with rendering as a subset of job and has the API docs available [here](https://docs.astro.build/en/reference/integrations-reference/)(checkout _Renderers_ under examples), but when they were just renderers I wrote [something](Renderers) that you might find a bit useful.

**Reply by matthewmueller** [reply]
Thanks for sharing your expertise @one-aalam! I'll definitely look into Astro when we start supporting other renderers.

**Comment by donuts-are-good** [comment]
Not interested in React and any JS trash (I mean it. fight me.), just bare HTML. Does bare HTML belong in this issue or should it be a new one?

**Reply by matthewmueller** [reply]
There will be no fighting here 😇

I just split HTML support into a new discussion: https://github.com/livebud/bud/discussions/40

**Reply by donuts-are-good** [reply]
Thanks

**Comment by austincollinpena** [comment]
Is React somewhere on the timeline?

**Reply by matthewmueller** [reply]
Not yet! I'm still working out the direction to take view support (see [this comment](https://github.com/livebud/bud/discussions/188#discussioncomment-3152348) for details).

**Comment by talesmgodois** [comment]
Maybe we can start thinking about create plugins. 
For example, I would like to use svelte with typescript. However, how my fast refresh would work. Right now my idea is to use webpack or something similar to make this, but It would be really nice to have something related to bud framework.
